Pokedex/Networking/PokemonClient.swift

```diff
@@ -0,0 +1,110 @@
+            guard let onePokemon = onePokemon else {
+                completion(nil, error)
+                return
+            }
+            completion(onePokemon, nil)
+        }
+    }
+    
+    func fetchPokemonSprite(withSprite sprite: String, using session: URLSession = URLSession.shared, completion: @escaping (Data?, Error?) -> Void) {
+
+        guard let pokemonSpriteURL = self.urlPokemonSprite(withSprite: sprite) else { return }
+        fetchImageData(from: pokemonSpriteURL, using: session) { (pokemonSprite: Data?, error: Error?) in
+            guard let pokemonSprite = pokemonSprite else {
+                completion(nil, error)
+                return
+            }
+
+            completion(pokemonSprite, nil)
+        }
+    }
+    
+    private func fetch<T: Decodable> (from url: URL, using session: URLSession = URLSession.shared, completion: @escaping (T?, Error?) -> Void) {
+        session.dataTask(with: url) {(data, response, error) in
+            if let error = error {
+                completion(nil, error)
+                return
+            }
+            
+            guard let data = data else {
+                completion(nil, NSError(domain: "", code: -1, userInfo: nil))
+                return
+            }
+            
+            do {
+                let jsonDecoder = JSONDecoder()
+                let decodedObject = try jsonDecoder.decode(T.self, from: data)
+                completion(decodedObject, nil)
+            } catch {
+                completion(nil, error)
+            }
+        }.resume()
+    }
+    
+    private func fetchImageData(from url: URL, using session: URLSession = URLSession.shared, completion: @escaping (Data?, Error?) -> Void) {
+        session.dataTask(with: url) {(data, response, error) in
+            if let error = error {
+                completion(nil, error)
+                return
+            }
+            
+            guard let data = data else {
+                completion(nil, NSError(domain: "", code: -1, userInfo: nil))
+                return
+            }
+            completion(data, nil)
+        }.resume()
+    }
+    
+    private func urlAllPokemons() -> URL {
+        var allPokemonsURL = baseURL
+        allPokemonsURL.appendPathComponent("pokemon")
+        let allPokemonsURLComp = NSURLComponents(url: allPokemonsURL, resolvingAgainstBaseURL: true)!
+        allPokemonsURLComp.queryItems = [URLQueryItem(name: "limit", value: "100"), URLQueryItem(name: "offset", value: "0")]
+        return allPokemonsURLComp.url!
+    }
+    
+    private func urlOnePokemon(forPokemon pokemonName: String) -> URL {
+        var pokemonURL = baseURL
+        pokemonURL.appendPathComponent("pokemon")
+        pokemonURL.appendPathComponent(pokemonName)
+        return pokemonURL
+    }
+    
+    private func urlPokemonSprite(withSprite sprite: String) -> URL? {
+        guard let spriteURL = URL(string: sprite) else { return nil }
+        return spriteURL
+    }
```

Comment:
You have methods for fetching and for building URLs. This is a perfect case for `Single responsibility principle` (S in SOLID). Extract the methods for building URLs in a `PokemonPathBuilder` struct/class so it can ultimately be tested. It will be a simple exercice for writing your first unit tests !

Reply:
An example of what I've done recently :

```swift
struct EpisodePathBuilder {
    // These are String constants
    let cloudFrontRoot = // ...
    let version = // ...
    let episodes = // ...
    let episodesList = // ...
    let scenesChoicesTime = // ...
    let scenesTree = // ...
    let hls = // ...
    let scenes = // ...

    func buildEpisodeFolderPath(for episodeKey: String) -> String {
        "\(cloudFrontRoot)/\(version)/\(episodes)/\(episodeKey)"
    }

    func buildScenesTreePath(
        for episodeKey: String,
        locale: String
    ) -> String {
        "\(buildEpisodeFolderPath(for: episodeKey))/\(scenesTree)/\(scenesTree)-\(locale).json"
    }

    func buildChoicesTimesPath(for episodeKey: String) -> String {
        "\(buildEpisodeFolderPath(for: episodeKey))/\(scenesChoicesTime)"
    }
}
```